sdk/src/main/scala/com/horizen/account/receipt/EthereumConsensusDataReceipt.scala

```diff
@@ -188,54 +225,57 @@ object EthereumConsensusDataReceipt{
-                                bin.add(b[:], buf)
-                            }
-                        }
-                    }
-                    return bin
-                }
-
-                // add is internal version of Add, which takes a scratch buffer for reuse (needs to be at least 6 bytes)
-                func (b *Bloom) add(d []byte, buf []byte) {
-                    i1, v1, i2, v2, i3, v3 := bloomValues(d, buf)
-                    b[i1] |= v1
-                    b[i2] |= v2
-                    b[i3] |= v3
-                }
-
-                // bloomValues returns the bytes (index-value pairs) to set for the given data
-                func bloomValues(data []byte, hashbuf []byte) (uint, byte, uint, byte, uint, byte) {
-                    sha := hasherPool.Get().(crypto.KeccakState)
-                    sha.Reset()
-                    sha.Write(data)
-                    sha.Read(hashbuf)
-                    hasherPool.Put(sha)
-                    // The actual bits to flip
-                    v1 := byte(1 << (hashbuf[1] & 0x7))
-                    v2 := byte(1 << (hashbuf[3] & 0x7))
-                    v3 := byte(1 << (hashbuf[5] & 0x7))
-                    // The indices for the bytes to OR in
-                    i1 := BloomByteLength - uint((binary.BigEndian.Uint16(hashbuf)&0x7ff)>>3) - 1
-                    i2 := BloomByteLength - uint((binary.BigEndian.Uint16(hashbuf[2:])&0x7ff)>>3) - 1
-                    i3 := BloomByteLength - uint((binary.BigEndian.Uint16(hashbuf[4:])&0x7ff)>>3) - 1
-
-                    return i1, v1, i2, v2, i3, v3
-                }
-
-             */
-    // TODO shall we use libevm implementation?
-    new Array[Byte](256)
+  def bytesToShort(byte1: Byte, byte2: Byte): Short = {
+    ByteBuffer
+      .allocate(2)
+      .order(ByteOrder.BIG_ENDIAN)
+      .put(byte1)
+      .put(byte2)
+      .getShort(0)
+  }
+
+  def getBloomFilterValues(data: Array[Byte]) = {
+    // Copy first 3 pairs of bytes to hashBuffer
+    val hashBuffer = Keccak256.hash(data)
+
+    // Calculate byte values
+    val v1 = 1 << (hashBuffer(1) & 0x7)
+    val v2 = 1 << (hashBuffer(3) & 0x7)
+    val v3 = 1 << (hashBuffer(5) & 0x7)
+
+    // Calculate byte indexes
+    val i1 =
+      256 - ((bytesToShort(hashBuffer(0), hashBuffer(1)) & 0x7ff) >> 3) - 1
+    val i2 =
+      256 - ((bytesToShort(hashBuffer(2), hashBuffer(3)) & 0x7ff) >> 3) - 1
+    val i3 =
+      256 - ((bytesToShort(hashBuffer(4), hashBuffer(5)) & 0x7ff) >> 3) - 1
+
+    (i1, v1, i2, v2, i3, v3)
+  }
+
+  def addDataToBloomFilter(
+      bloomFilter: Array[Byte],
+      data: Array[Byte]
+  ): Array[Byte] = {
+    val (i1, v1, i2, v2, i3, v3) = getBloomFilterValues(data)
+
+    bloomFilter(i1) = (bloomFilter(i1) | v1).toByte
+    bloomFilter(i2) = (bloomFilter(i2) | v2).toByte
+    bloomFilter(i3) = (bloomFilter(i3) | v3).toByte
+
+    bloomFilter
+  }
+
+  def createBloomFilter(evmLog: Seq[EvmLog]): Array[Byte] = {
```

Comment:
I suggest to create a separate class `BloomFilter` that will encapsulate all the logic of bloom filter inside:
```
class BloomFilter {
  private val bloomFilter: Array[Byte] = new Array[Byte](BLOOM_FILTER_LENGTH)

  def add(data: Array[Byte]): Unit = {
    ///... like addDataToBloomFilter
  }
  def getBloomFilter: Array[Byte] = { /* copy of bloomFilter array*/ }
  private def bytesToShort(byte1: Byte, byte2: Byte): Short = { ... }

  private def getBloomFilterValues(data: Array[Byte]) = { ... }
}
object BloomFilter {
  def apply(evmLog: Seq[EvmLog]): BloomFilter =  { /* create filter and add all the logs */ }
  def apply(evmLog: Seq[EthereumReceipt]): BloomFilter = { /* create filter and add all the receipts' logs */ }
  
  val BLOOM_FILTER_LENGTH: Int = 256
}
```

Reply:
Yes, this please. I just have two things to add:
- Also add `Test(data: Array[Byte]): Boolean` to check presence of data in the bloom filter. See geth implementation below.
- We should also be able to reconstruct a BloomFilter instance from the resulting data like `new BloomFilter(bloomFilterArray)` to be able to use the Test method.
```
// Test checks if the given topic is present in the bloom filter
func (b Bloom) Test(topic []byte) bool {
	i1, v1, i2, v2, i3, v3 := bloomValues(topic, make([]byte, 6))
	return v1 == v1&b[i1] &&
		v2 == v2&b[i2] &&
		v3 == v3&b[i3]
}
```

Reply:
> I suggest to create a separate class `BloomFilter` that will encapsulate all the logic of bloom filter inside:
> 
> ```
> class BloomFilter {
>   private val bloomFilter: Array[Byte] = new Array[Byte](BLOOM_FILTER_LENGTH)
> 
>   def add(data: Array[Byte]): Unit = {
>     ///... like addDataToBloomFilter
>   }
>   def getBloomFilter: Array[Byte] = { /* copy of bloomFilter array*/ }
>   private def bytesToShort(byte1: Byte, byte2: Byte): Short = { ... }
> 
>   private def getBloomFilterValues(data: Array[Byte]) = { ... }
> }
> object BloomFilter {
>   def apply(evmLog: Seq[EvmLog]): BloomFilter =  { /* create filter and add all the logs */ }
>   def apply(evmLog: Seq[EthereumReceipt]): BloomFilter = { /* create filter and add all the receipts' logs */ }
>   
>   val BLOOM_FILTER_LENGTH: Int = 256
> }
> ```

I've tried creating two different apply methods as you've suggested, but Scala doesn't allow me to create them because both arguments are of type Seq (I guess that Scala only checks if it is Seq, not Seq[Type]) so it doesn't differentiate them. I also tried creating two auxiliary constructors based on your approach, I get the same problem.

Reply:
Ah yes, that doesn't work, your suggestion is ok.